probe-rs/src/architecture/arm/traits.rs

```diff
@@ -45,21 +45,66 @@ pub enum DpAddress {
     Multidrop(u32),
 }
 
+/// Access port v2 address
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub enum ApV2Address {
+    /// Last node of an APv2 address
+    Leaf(u32),
+    /// Non-terminal node of an APv2 address
+    Node(u32, Box<ApV2Address>),
+}
+
+impl std::fmt::Display for ApV2Address {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            ApV2Address::Leaf(v) => write!(f, "{}", v),
+            ApV2Address::Node(v, r) => write!(f, "{}.{}", v, r),
+        }
+    }
+}
+
+/// Access port address
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub enum ApAddress {
+    /// Access port v1 address
+    V1(u8),
+    /// Access Port v2
+    V2(ApV2Address),
+}
+
+impl std::fmt::Display for ApAddress {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            ApAddress::V1(v) => write!(f, "V1({})", v),
+            ApAddress::V2(v) => write!(f, "V2({})", v),
+        }
+    }
+}
+
 /// Access port address.
-#[derive(Debug, PartialEq, Eq, Clone, Copy)]
-pub struct ApAddress {
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct FullyQualifiedApAddress {
```

Comment:
What would be a not fully qualified AP address? Could you add comment to explain the difference?

Reply:
The ApAddress without the DP address.

I’m notoriously bad at this, so feel free to suggest alternatives.